File: java/src/main/java/icepackai/ns3helper.java
package icepackai;

import java.util.*;

import icepackai.NS3.Ns3Tbfvuwtge2Iq.InternalDimension.eMeasurementUnit;
import icepackai.CVRP.CvrpJkfdoctmp51N.SolutionResponse;
import icepackai.NS3.*;

import dnl.utils.text.table.TextTable;

// some helper functions for building ns3 models.
public class ns3helper {
  public static Ns3Tbfvuwtge2Iq.DimensionConfiguration makeDistanceTimeUserDims(String userDim) {
    Ns3Tbfvuwtge2Iq.DimensionConfiguration.Builder dimBuilder =
        Ns3Tbfvuwtge2Iq.DimensionConfiguration.newBuilder();
    dimBuilder.setDistanceConfig(Ns3Tbfvuwtge2Iq.InternalDimension.newBuilder()
                                     .setId("distance")
                                     .setMeasurementUnit(eMeasurementUnit.KILOMETRES)
                                     .build());
    dimBuilder.setTimeConfig(Ns3Tbfvuwtge2Iq.InternalDimension.newBuilder()
                                 .setId("time")
                                 .setMeasurementUnit(eMeasurementUnit.HOURS)
                                 .build());
    if (userDim != "") {
      dimBuilder.addUserDimensions(
          Ns3Tbfvuwtge2Iq.UserDimension.newBuilder().setId(userDim).setUnits("unknown").build());
    }
    return dimBuilder.build();
  }

  public static Ns3Tbfvuwtge2Iq.Node makeNode(String id, float longitude, float latitude) {
    Ns3Tbfvuwtge2Iq.Node.Builder n = Ns3Tbfvuwtge2Iq.Node.newBuilder();
    n.setId(id).setGeocode(
        Ns3Tbfvuwtge2Iq.Geocode.newBuilder().setLongitude(longitude).setLatitude(latitude).build());
    return n.build();
  }

  public static List<Ns3Tbfvuwtge2Iq.Node> makeNodes(List<dataRow> rows) {
    List<Ns3Tbfvuwtge2Iq.Node> nodes = new ArrayList<Ns3Tbfvuwtge2Iq.Node>();
    for (int i = 0; i < rows.size(); i++) {
      dataRow loc = rows.get(i);
      nodes.add(makeNode(loc.id, loc.X, loc.Y));
    }
    return nodes;
  }

  public static Ns3Tbfvuwtge2Iq.DimensionRange make_dimension_range(
      String dimid, float min, float max) {
    Ns3Tbfvuwtge2Iq.DimensionRange.Builder dr = Ns3Tbfvuwtge2Iq.DimensionRange.newBuilder();
    return dr.setDimensionId(dimid).setMaxRange(max).setMinRange(min).setFlowPenalty(1e6f).build();
  }

  public static Ns3Tbfvuwtge2Iq.UnitDimensionCost make_udc(
      String dimid, float coef, float costperunit) {
    Ns3Tbfvuwtge2Iq.UnitDimensionCost.Builder udc = Ns3Tbfvuwtge2Iq.UnitDimensionCost.newBuilder();
    return udc.addDimensionIds(dimid)
        .addDimensionCoefficients(coef)
        .setCostPerUnit(costperunit)
        .build();
  }

  public static Ns3Tbfvuwtge2Iq.LaneRate make_lane_rate_distance(
      String src, String dest, float costperkm) {
    Ns3Tbfvuwtge2Iq.LaneRate.Builder lr = Ns3Tbfvuwtge2Iq.LaneRate.newBuilder();
    return lr.setId("lr:" + src + "->" + dest)
        .setSource(src)
        .setDestination(dest)
        .addUnitDimensionCosts(make_udc("distance", 1.0f, costperkm))
        .build();
  }

  public static Ns3Tbfvuwtge2Iq.LaneRate make_lane_rate_distance_weight(
      String src, String dest, float costperkm, String weightdim, float costperunit) {
    Ns3Tbfvuwtge2Iq.LaneRate.Builder lr = Ns3Tbfvuwtge2Iq.LaneRate.newBuilder();
    lr = lr.setId("lr:" + src + "->" + dest)
             .setSource(src)
             .setDestination(dest)
             .addUnitDimensionCosts(make_udc(weightdim, 1.0f, costperunit));
    if (costperkm != 0.0f) {
      return lr.addUnitDimensionCosts(make_udc("distance", 1.0f, costperkm)).build();
    }
    return lr.build();
  }

  public static Ns3Tbfvuwtge2Iq.ProductGroup make_simple_product_group(String product) {
    Ns3Tbfvuwtge2Iq.ProductGroup.Builder pg = Ns3Tbfvuwtge2Iq.ProductGroup.newBuilder();
    return pg.setProductId(product).setProductGroupId(product).build();
  }

  public static Ns3Tbfvuwtge2Iq.CostModel make_cost_model_distance(String src, float costperkm) {
    Ns3Tbfvuwtge2Iq.CostModel.Builder cm = Ns3Tbfvuwtge2Iq.CostModel.newBuilder();
    return cm.setId("costmodel: " + src + ":Beer")
        .setSource(src)
        .addProductGroupIds("Beer")
        .addUnitDimensionCost(make_udc("distance", 1.0f, costperkm))
        .build();
  }

  public static void printSolution(Ns3Tbfvuwtge2Iq.SolutionResponse solution) {
    System.out.println("Assignment table:");
    String[] columnNames = new String[] {"Lane Rate", "Cost Model", "Source", "Destination",
        "Product", "Amount", "Cost", "Distance", "Duration"};
    List<Object[]> tabData = new ArrayList<Object[]>();

    Integer maxPrint = 0;
    for (Ns3Tbfvuwtge2Iq.SolutionResponse.Assignment r : solution.getAssignmentsList()) {
      tabData.add(
          new Object[] {r.getLaneRateId(), r.getCostModelId(), r.getSource(), r.getDestination(),
              r.getProductId(), r.getAmount(), r.getCost(), r.getDistance(), r.getDuration()});
      maxPrint++;
      if (maxPrint > 10) {
        break;
      }
    }
    printTable(tabData, columnNames);
    if (maxPrint > 10) {
      System.out.println("Ouputput truncated..." + (solution.getAssignmentsCount() - maxPrint)
          + " items skipped.");
    }

    columnNames = new String[] {"NodeId", "InFlow", "OutFlow", "FixedCost", "FlowCost",
        "ProductFixedCost", "ProductFlowCost", "P-Amount", "P-Cost", "P-Penalty", "C-Amount",
        "C-Cost", "C-Penalty"};
    tabData.clear();
    maxPrint = 0;
    for (Ns3Tbfvuwtge2Iq.SolutionResponse.NodeFlow nf : solution.getNodeFlowsList()) {
      tabData.add(new Object[] {nf.getNodeId(), nf.getInFlow(), nf.getOutFlow(), nf.getFixedCost(),
          nf.getFlowCost(), nf.getProductFixedCost(), nf.getProductFlowCost(),
          nf.getProductionAmount(), nf.getProductionCost(), nf.getProductionPenalty(),
          nf.getConsumptionAmount(), nf.getConsumptionCost(), nf.getConsumptionPenalty()});
      maxPrint++;
      if (maxPrint > 10) {
        break;
      }
    }
    System.out.println("\nNode Flow table:");
    printTable(tabData, columnNames);
    if (maxPrint > 10) {
      System.out.println("Ouputput truncated..." + (solution.getAssignmentsCount() - maxPrint)
          + " items skipped.");
    }

    columnNames = new String[] {"NodeId", "ProductId", "InFlow", "OutFlow", "FixedCost", "FlowCost",
        "P-Amount", "P-Cost", "P-Penalty", "C-Amount", "C-Cost", "C-Penalty"};
    tabData.clear();
    maxPrint = 0;
    for (Ns3Tbfvuwtge2Iq.SolutionResponse.NodeProductFlow nf : solution.getNodeProductFlowsList()) {
      tabData.add(new Object[] {nf.getNodeId(), nf.getInFlow(), nf.getProductId(), nf.getOutFlow(),
          nf.getFixedCost(), nf.getFlowCost(), nf.getProductionAmount(), nf.getProductionCost(),
          nf.getProductionPenalty(), nf.getConsumptionAmount(), nf.getConsumptionCost(),
          nf.getConsumptionPenalty()});
      maxPrint++;
      if (maxPrint > 10) {
        break;
      }
    }
    System.out.println("\nNode Product Flow table:");
    printTable(tabData, columnNames);
    if (maxPrint > 10) {
      System.out.println("Ouputput truncated..." + (solution.getAssignmentsCount() - maxPrint)
          + " items skipped.");
    }

    // then just for the curious reader. Each of the geometries is available in the payload. The
    // route defines the sequence of geometry segments that should be joined to form the linestring
    // for a particular route.
    List<List<double[]>> routes = new ArrayList<List<double[]>>();
    for (Ns3Tbfvuwtge2Iq.SolutionResponse.Route r : solution.getRoutesList()) {
      List<double[]> linestring = new ArrayList<double[]>();
      for (int index : r.getGeometrySequenceList()) {
        Ns3Tbfvuwtge2Iq.SolutionResponse.GeometrySequence gs = solution.getGeometrySequence(index);
        for (int i = 0; i < gs.getXCount(); i++) {
          linestring.add(new double[] {gs.getX(i), gs.getY(i)});
        }
      }
      routes.add(linestring);
    }
  }

  private static void printTable(List<Object[]> tabData, String[] columnNames) {
    Object[][] tab = new Object[tabData.size()][];
    for (int i = 0; i < tabData.size(); i++) {
      tab[i] = tabData.get(i);
    }
    TextTable tt = new TextTable(columnNames, tab);
    tt.printTable();
  }
}

File: csharp/ns3helper.cs
using System.Collections.Generic;
using System;

class ns3helper{

  public static Ns3.DimensionConfiguration make_distance_time_user_dimensions(string userDimension){

    var dimconfig = new Ns3.DimensionConfiguration();
    dimconfig.distanceConfig = new Ns3.InternalDimension{
      Id = "distance",
      measurementUnit =  Ns3.InternalDimension.eMeasurementUnit.Kilometres,
    };

    dimconfig.timeConfig = new Ns3.InternalDimension{
      Id = "time",
      measurementUnit = Ns3.InternalDimension.eMeasurementUnit.Hours,
    };
    if(userDimension != ""){
      dimconfig.userDimensions.Add(new Ns3.UserDimension{
        Id = userDimension,
        Units = "unknown"
      });
    }

    return dimconfig;
  }

  public static Ns3.Node make_node(string id, float longitude, float latitude){
    var node = new Ns3.Node{
      Id = id,
      Geocode = new Ns3.Geocode{
        Longitude = longitude,
        Latitude = latitude,
      }
    };
    return node;
  }

  public static List<Ns3.Node> make_nodes(List<dataRow> rows ){
      List<Ns3.Node> nodes = new List<Ns3.Node>();
      foreach( var n in rows){
        nodes.Add(make_node(n.id, n.X, n.Y));
      }
      return nodes;
  }

  public static Ns3.DimensionRange make_dimension_range(string dimid, float min, float max){
    var dr = new Ns3.DimensionRange{
      dimensionId = dimid,
      maxRange = max,
      minRange = min,
      flowPenalty = 1e6f,
    };
    return dr;
  }

  public static Ns3.UnitDimensionCost make_udc(string dimid, float coef, float costperunit){
    var udc = new Ns3.UnitDimensionCost{
      dimensionIds = {dimid},
      dimensionCoefficients = new float[]{coef},
      costPerUnit = costperunit
    };
    return udc;
  } 

  public static Ns3.LaneRate make_lane_rate_distance(string src, string dest, float costperkm){
    var lr = new Ns3.LaneRate{
      Id = ("lr:" + src + "->" + dest),
      Source = src,
      Destination = dest,
      unitDimensionCosts = {make_udc("distance", 1.0f, costperkm)}
    };
    return lr;
  }

  public static Ns3.LaneRate make_lane_rate_distance_weight(string src, string dest, float costperkm, string weightdim, float costperunit){
    var lr = new Ns3.LaneRate{
      Id = ("lr:" + src + "->" + dest),
      Source = src,
      Destination = dest,
      unitDimensionCosts = {make_udc(weightdim, 1.0f, costperunit)}
    };
    if(costperkm != 0.0f){
      lr.unitDimensionCosts.Add(make_udc("distance", 1.0f, costperkm));
    }
    return lr;
  }

  public static Ns3.ProductGroup make_single_product_group(string product){
    var pg = new Ns3.ProductGroup{
      productGroupId = product,
      productId = product
    };
    return pg;
  }

  public static Ns3.CostModel make_cost_model_distance(string src, float costperkm){
    var cm = new Ns3.CostModel{
      Id = ("costmodel: " + src + ":Beer"),
      Source = src,
      productGroupIds = {"Beer"},
      unitDimensionCosts = {make_udc("distance", 1.0f, costperkm)}
    };
    return cm;
  }

  public static void printSolution(Ns3.SolutionResponse resp, Ns3.SolveRequest sr){
    // we're just going to show the first 5 items from each table that one can construct here.
    // we also give an illustration of how to construct the geometries in this context.

    //assignment table.
    int[] maxchar = new int[]{ "Lane Rate".Length, "Cost Model".Length, "Source".Length,
    "Destination".Length, "Product".Length, "Amount".Length, "Cost".Length, "Distance".Length, "Duration".Length };
    foreach(var a in resp.Assignments){
      maxchar[0] = Math.Max(maxchar[0], a.laneRateId.Length);
      maxchar[1] = Math.Max(maxchar[1], a.costModelId.Length);
      maxchar[2] = Math.Max(maxchar[2], a.Source.Length);
      maxchar[3] = Math.Max(maxchar[3], a.Destination.Length);
      maxchar[4] = Math.Max(maxchar[4], a.productId.Length);
      //a.Amount
      //a.Cost
      //a.Distance
      //a.Duration
    }
    Console.WriteLine("Assignments:");
    string formatLine = "|{0,-" + maxchar[0] + "}|{1,-" + maxchar[1] + "}|{2,-" + maxchar[2] + "}|{3,-" +
     maxchar[3] + "}|{4,-" + maxchar[4] + "}|{5,10}|{6,10}|{7,10}|{8,10}|";
    Console.WriteLine(String.Format(formatLine, "Lane Rate", "Cost Model", "Source", "Destination", "Product", "Amount", "Cost", "Distance", "Duration"));
    int maxprint = 0;
    foreach(var a in resp.Assignments){
      Console.WriteLine(String.Format(formatLine, a.laneRateId, a.costModelId, a.Source, a.Destination, a.productId,
                        string.Format("{0:0.00}", a.Amount),
                        string.Format("{0:0.00}", a.Cost),
                        string.Format("{0:0.00}", a.Distance),
                        string.Format("{0:0.00}", a.Duration)));
                        maxprint++;
      if(maxprint > 10){
        Console.WriteLine("table truncated...");
        maxprint = 0;
        break;
      }
    }
    Console.WriteLine("");

    Console.WriteLine("Node Flow:");
    maxchar[0] = "NodeId".Length;

    foreach(var nf in resp.nodeFlows){
        maxchar[0] = Math.Max(maxchar[0], nf.nodeId.Length);
    }

    formatLine = "|{0,-" + maxchar[0] + "}|{1,10}|{2,10}|{3,10}|{4,10}|{5," + "ProductFixedCost".Length+ "}|{6," + 
    "ProductFlowCost".Length +"}|{7,10}|{8,10}|{9,10}|{10,10}|{11,10}|{12,10}|";
    Console.WriteLine(String.Format(formatLine, "NodeId", "InFlow", "OutFlow", "FixedCost", "FlowCost",  
            "ProductFixedCost", "ProductFlowCost", "P-Amount", "P-Cost", "P-Penalty", "C-Amount", "C-Cost", "C-Penalty"));
    
    foreach(var nf in resp.nodeFlows){
      Console.WriteLine(String.Format(formatLine, nf.nodeId, 
                        string.Format("{0:0.00}", nf.inFlow),
                        string.Format("{0:0.00}", nf.outFlow),
                        string.Format("{0:0.00}", nf.fixedCost),
                        string.Format("{0:0.00}", nf.flowCost),
                        string.Format("{0:0.00}", nf.productFixedCost),
                        string.Format("{0:0.00}", nf.productFlowCost),
                        string.Format("{0:0.00}", nf.productionAmount),
                        string.Format("{0:0.00}", nf.productionCost),
                        string.Format("{0:0.00}", nf.productionPenalty),
                        string.Format("{0:0.00}", nf.consumptionAmount),
                        string.Format("{0:0.00}", nf.consumptionCost),
                        string.Format("{0:0.00}", nf.consumptionPenalty)));
      maxprint++;
      if(maxprint > 10){
        Console.WriteLine("table truncated...");
        maxprint = 0;
        break;
      }
    }
    Console.WriteLine("");

    Console.WriteLine("Node Product Flow:");
    maxchar[0] = "NodeId".Length;
    maxchar[1] = "ProductId".Length;

    foreach(var nf in resp.nodeFlows){
        maxchar[0] = Math.Max(maxchar[0], nf.nodeId.Length);
    }

    formatLine = "|{0,-" + maxchar[0] + "}|{1,-" + maxchar[1] +"}|{2,10}|{3,10}|{4,10}|{5,10}|{6,10}|{7,10}|{8,10}|{9,10}|{10,10}|{11,10}|";
    Console.WriteLine(String.Format(formatLine, "NodeId", "ProductId", "InFlow", "OutFlow", "FixedCost", "FlowCost",  
            "P-Amount", "P-Cost", "P-Penalty", "C-Amount", "C-Cost", "C-Penalty"));
    
    foreach(var nf in resp.nodeProductFlows){
      Console.WriteLine(String.Format(formatLine, nf.nodeId, nf.productId, 
                        string.Format("{0:0.00}", nf.inFlow),
                        string.Format("{0:0.00}", nf.outFlow),
                        string.Format("{0:0.00}", nf.fixedCost),
                        string.Format("{0:0.00}", nf.flowCost),
                        string.Format("{0:0.00}", nf.productionAmount),
                        string.Format("{0:0.00}", nf.productionCost),
                        string.Format("{0:0.00}", nf.productionPenalty),
                        string.Format("{0:0.00}", nf.consumptionAmount),
                        string.Format("{0:0.00}", nf.consumptionCost),
                        string.Format("{0:0.00}", nf.consumptionPenalty)));
      maxprint++;
      if(maxprint > 10){
        Console.WriteLine("table truncated...");
        maxprint = 0;
        break;
      }
    }


    // lets unpack the geometries for each of the "routes"
    List<lineString> routeData = new List<lineString>();
    foreach(var r in resp.Routes){
      lineString  ls = new lineString();
      foreach(var i in r.geometrySequences){
         for(int j = 0; j < resp.geometrySequences[i].X.Length; j++){
           ls.Add(new double[2]{resp.geometrySequences[i].X[j], resp.geometrySequences[i].Y[j]});
         }
      }
      routeData.Add(ls);
    }
    var cplot = new consolePlot(routeData);

  }

}